p4include/psa.p4

```diff
@@ -25,403 +24,518 @@ limitations under the License.
-
-  /// Clone operation: create a copy of the packet with additional
-  ///                  data and send it to the specified port.
-  ///
-  /// The PSA specifies four types of cloning, with the packet sourced
-  /// from different points in the pipeline and sent back to ingress
-  /// or to the buffering queue in the egress (@see CloneMethod_t).
-  ///
-  /// @param clone_method  The type of cloning.
-  /// @param port          The port to send the cloned packet to.
-  /// @param data additional header data attached to the packet
-  void clone<T>  (in CloneMethod_t clone_method, in PortId_t port, in T data);
-
-
-  /// Resubmit operation: send a packet to the ingress port with
-  ///                     additional data appended
-  ///
-  /// This operation is intended for recursive packet processing.
-  ///
-  /// @param data A header definition that can be added to the set of
-  ///             packet headers.
-  /// @param port The input port at which the packet will be resubmitted.
-  void resubmit<T>(in T data, in PortId_t port);
-
-  /// Recirculate operation: send a post deparse packet to the ingress
-  ///                        port with additional data appended
-  ///
-  /// This operation is intended for recursive packet processing.
-  ///
-  /// @param data A header definition that can be added to the set of
-  ///             packet headers.
-  /// @param port The input port at which the packet will be resubmitted.
-  void recirculate<T>(in T data, in PortId_t port);
-
-  /// Truncate operation
-  ///
-  /// Truncate the outgoing packet to the specified length
-  ///
-  /// @param length packet length
-  void truncate(in bit<32> length);
+extern BufferingQueueingEngine {
+    BufferingQueueingEngine();
+
+  // BufferingQueueingEngine(); /// No constructor. BQE is instantiated
+                                /// by the architecture.
 
-  /*
-  @ControlPlaneAPI
-  {
-  // TBD
-  }
-  */
 }
 
-/* --------------------------------------------------------------------- */
-/*
- * Hashes
- */
+// BEGIN:Clone_extern
+extern clone_out {
+  /// Write @hdr into the ingress/egress clone engine.
+  /// @T can be a header type, a header stack, a header union, or a struct
+  /// containing fields with such types.
+  void emit<T>(in T hdr);
+}
+// END:Clone_extern
 
-/// Supported hash algorithms
-enum HashAlgorithm {
-  crc32,
-  crc32_custom,
-  crc16,
-  crc16_custom,
-  random,
-  identity
-}
-
-/// Hash function
-///
-/// @param Algo The algorithm to use for computation (@see HashAlgorithm).
-/// @param O    The type of the return value of the hash.
-///
-/// Example usage:
-/// parser P() {
-///   Hash<crc16, bit<56>> h();
-///   bit<56> hash_value = h.getHash(16, buffer, 100);
-///   ...
-/// }
-extern Hash<Algo, O> {
-  /// Constructor
-  Hash();
 
-  /// compute the hash for data
-  O getHash<T, D, M>(in T base, in D data, in M max);
+// BEGIN:Resubmit_extern
+extern resubmit {
+  /// Write @hdr into the ingress packet buffer.
```

Comment:
Do you need emitPayload() or is it always happening?

Reply:
As mentioned above, resubmit extern will be deleted soon.